controller/controller.py

```diff
@@ -0,0 +1,274 @@
+    from vehicle.vehicle_control import InputChannel, Relay, BACKWARD_CAM_INDICES
+
+TRANSLATION_SENSITIVITY = 0.5
+ROTATIONAL_SENSITIVITY = 0.5
+
+FRONT_CAMERA_INDEX = 0
+BOTTOM_CAMERA_INDEX = 1
+BACK_CAMERA_INDEX = 2
+
+
+def apply_curve(value, exponential):
+    value = min(max(value, -1), 1)  # clamp the joystick input from -1 to 1
+    return math.copysign(abs(value) ** exponential, value)
+
+
+class Controller:
+    class JoystickAxis(Enum):
+        pass
+
+    class Trigger(Enum):
+        pass
+
+    class Button(Enum):
+        pass
+
+    def __init__(self, get_big_video_index, joystick_curve_exponential=2, trigger_curve_exponential=2):
+        self.MAX_JOY_VAL = None
+        self.MAX_TRIG_VAL = None
+
+        self.switch_camera_callbacks = []
+        self.toggle_relay_callbacks = []
+
+        self.get_big_video_index = get_big_video_index
+
+        self.joystick_curve_exponential = joystick_curve_exponential
+        self.trigger_curve_exponential = trigger_curve_exponential
+
+        self.joystick_axes: t.Dict[Controller.JoystickAxis, float] = {}
+        self.triggers: t.Dict[Controller.Trigger, float] = {}
+        self.buttons: t.Dict[Controller.Button, bool] = {}
+
+        self.joystick_axis_codes = [axis.value for axis in self.JoystickAxis]
+        self.trigger_codes = [trigger.value for trigger in self.Trigger]
+        self.button_codes = [button.value for button in self.Button]
+
+        self._monitor_thread = threading.Thread(target=self._monitor_controller, args=())
+        self._monitor_thread.daemon = True
+
+    def start_monitoring(self):
+        self._monitor_thread.start()
+
+    def _monitor_controller(self):
+        while True:
+            events = get_gamepad()
+            for event in events:
+                self._handle_event(event)
+
+    def _handle_event(self, event):
+        code = event.code
+        if "SYN" in code:
+            return
+        if code in self.joystick_axis_codes:
+            self.joystick_axes[self.JoystickAxis(
+                code)] = event.state / self.MAX_JOY_VAL  # normalize between -1 and 1
+        elif code in self.trigger_codes:
+            self.triggers[self.Trigger(code)] = event.state / self.MAX_TRIG_VAL
+        elif code in self.button_codes:
+            self.buttons[self.Button(code)] = bool(event.state)
+        else:
+            print(f"Unrecognized: {code}")
+
+        self.check_for_camera_change(event)
+        self.check_for_relay_toggle(event)
+
+    def get(self, control: t.Union[JoystickAxis, Trigger, Button]) -> t.Union[float, bool]:
+        if isinstance(control, self.JoystickAxis):
+            return apply_curve(self.joystick_axes.get(control, 0), self.joystick_curve_exponential)
+        if isinstance(control, self.Trigger):
+            return apply_curve(self.triggers.get(control, 0), self.trigger_curve_exponential)
+        if isinstance(control, self.Button):
+            return self.buttons.get(control, False)
```

Comment:
Wouldn't it be possible to implement this method separately under the `JoystickAxis` `Trigger` and `Button` classes so we don't have to do these `isinstance` checks?

Reply:
I don't see how a method of one of those enums would have access to the `joystick_axes`, `triggers`, or `buttons` variables. We could just split it into three methods but this is more convenient imo.

Reply:
Hmm ok. I forget nested classes work different in python from Java.